### DISCUSSION TITLE
Have you played around making executables with graalvm?

### DISCUSSION THREAD

**Discussion by ckipp01** [oyvindberg/tui-scala#37] [Q&A] [answered]
Do you by chance have any tips about producing an image via graalvm for a `tui-scala` project? For example, if you take any of your example, say `ListExample` and slap this on top:

```
//> using scala "3.3.0-RC3"
//> using lib "com.olvind.tui::tui:0.0.5"
```

 I was hoping I'd be able to do:

```
scala-cli package --native-image --graalvm-java-version 19 .
```

But when doing that I end up with:

```
❯ ./ListExample
Exception in thread "main" java.lang.UnsatisfiedLinkError: Native library libcrossterm.dylib (/libcrossterm.dylib) cannot be found on the classpath.
        at tui.crossterm.NativeLoader.loadPackaged(NativeLoader.java:29)
        at tui.crossterm.NativeLoader.load(NativeLoader.java:11)
        at tui.crossterm.CrosstermJni.<clinit>(CrosstermJni.java:9)
        at tui.withTerminal$.apply(withTerminal.scala:6)
        at ListExample$.main(project.scala:124)
        at ListExample.main(project.scala)
```

After you try to run the executable it creates. Due to crossterm, is there any special things or flags that need to be used when making an image this way?

**Comment by oyvindberg** [comment]
Yes, it does work! However, one manual step is needed, because graalvm has some trouble with the path structure i inherited from the SBT plugin.

I'll post details about the manual step once I'm on a computer, in some hours

**Comment by oyvindberg** [comment]
See #28 for now

**Reply by ckipp01** [reply]
ahhh ok that's helpful. I wonder how I can translate this to scala-cli 🤔 . Also sorry I didn't do a better job of searching the issues 😆 I glossed right over that when I looked.

**Reply by oyvindberg** [reply, marked as answer]
Ok, so looking at graalvm milestones it looks like they are aiming to finally release the new version in nine days, which means that I'll wait for that and then polish this experience afterwards. 

In the meantime what you need to do is this:

### Extract the native library you want to load from crossterm jar. 

```
$ jar -tf `cs fetch com.olvind.tui:crossterm:0.0.5`|grep native
native/arm64-darwin/libcrossterm.dylib
native/x86_64-darwin/libcrossterm.dylib
native/x86_64-linux/libcrossterm.so
native/x86_64-windows/crossterm.dll
META-INF/native-image/org.scala-lang/scala-lang/jni-config.json
```

### Copy the library for you platform so it loses the directory prefix

for instance `native/arm64-darwin/libcrossterm.dylib` should go into a `resources` directory with the name `libcrossterm.dylib`

(if you need to debug this, you'll find the logic for dropping the prefix in `https://github.com/oyvindberg/tui-scala/blob/master/crossterm/src/java/tui/crossterm/NativeLoader.java`. I indent to simplify this after the new graalvm release. Another option is to encode platform in the library filename instead of in folder names. I think that will also solve the issue.)


### Run native-image

The options I used for native-image are [here](https://github.com/oyvindberg/tui-scala/blob/master/scripts/src/scala/tui/scripts/GenNativeImage.scala#L31-L39)
```scala
    val plugin = new NativeImagePlugin(
      project = started.bloopProjects(demoProject),
      logger = started.logger,
      nativeImageOptions = List(
        "--verbose",
        "--no-fallback",
        "-H:+ReportExceptionStackTraces",
        "--initialize-at-build-time=scala.runtime.Statics$VM",
        "--initialize-at-build-time=scala.Symbol",
        "--initialize-at-build-time=scala.Symbol$",
        "--native-image-info",
        """-H:IncludeResources=libcrossterm.dylib""",
        "-H:-UseServiceLoaderFeature"
      ),
      jvmCommand = jvmCommand,
      env = sys.env.toList ++ List(("USE_NATIVE_IMAGE_JAVA_PLATFORM_MODULE_SYSTEM", "false"))
    )
```

**Reply by ckipp01** [reply]
lol so hacky, but totally working for me. Thanks so much! If anyone stumbles on this and needs a somewhat standalone version to do this you can use [this script](https://github.com/ckipp01/skan/blob/main/scripts/package-setup.scala). It _maybe_ will work on your machine. 😆 hopefully this will be easier when the new graal is out. Thanks again!

**Reply by oyvindberg** [reply]
Lol yeah. It was supposed to be seamless :)

I'll update here in nineish days with a better approach :)

**Comment by oyvindberg** [comment]
The graalvm release seems delayed until late june now. maybe time to find a better solution here in the interim. it should really just be some name standard graalvm is able to pick up - maybe replace `/` by `_` or something

**Reply by oyvindberg** [reply]
I think I fixed this yesterday, but something in the publishing code has bitrotted. will try to get out a new release soon so it can be tested properly

**Comment by oyvindberg** [comment]
I released 0.0.6 with the fix for this - try it out!

the bleep script for creating native images looks like this after the removal of all the workarounds https://github.com/oyvindberg/tui-scala/blob/master/scripts/src/scala/tui/scripts/GenNativeImage.scala

**Reply by ckipp01** [reply]
Alright, just tried this out and I think something is off. I haven't dug into it deeper yet, but it looks like _maybe_ since it's publishing on linux it's only including the linux file? For example on Mac I am now getting

```
Exception in thread "main" java.lang.UnsatisfiedLinkError: Native library libnative-x86_64-darwin-crossterm.dylib (/libnative-x86_64-darwin-crossterm.dylib) cannot be found on the classpath.
        at tui.crossterm.NativeLoader.loadPackaged(NativeLoader.java:20)
        at tui.crossterm.NativeLoader.load(NativeLoader.java:11)
        at tui.crossterm.CrosstermJni.<clinit>(CrosstermJni.java:9)
        at tui.withTerminal$.apply(withTerminal.scala:6)
        at skan.project$package$.run(project.scala:249)
        at skan.run.main(project.scala:19)
```

And when I look at the unpacked jar I see:

```
.rw-r--r--@ 4.8M ckipp  1 Jan  2010  libnative-x86_64-linux-crossterm.so
drwxr-xr-x@    - ckipp 30 Apr 18:53  META-INF
drwxr-xr-x@    - ckipp 
```

So I'd be expecting to see all of them there, not just the linux one right?

**Reply by oyvindberg** [reply]
Ok, thanks! I'll give it another look then.

**Reply by ckipp01** [reply]
Just confirming that the 0.0.7 release does indeed fix this for me. Thanks! Way easier to make an image now 😆

**Reply by oyvindberg** [reply]
Perfect! :)
